File: frontend/src/api/certificates.ts
import { apiClient } from './client'

export interface CertificateSummary {
  id: number
  serial_hex: string
  profile_code: string
  subject_dn: string
  sans: string[]
  status: string
  not_before: string
  not_after: string
  issued_via: string
  requested_by_user_id: number | null
}

export interface IssueCertificateResponse extends CertificateSummary {
  certificate_pem: string
  chain_pem: string
}

export interface RenewCertificateResponse extends IssueCertificateResponse {
  predecessor_superseded: boolean
}

export interface CertificateFilters {
  q?: string
  status?: string
  profile_code?: string
  issued_via?: string
  valid?: boolean
}

export const certificatesApi = {
  list: (filters?: CertificateFilters) => {
    const params = new URLSearchParams()
    if (filters?.q) params.set('q', filters.q)
    if (filters?.status) params.set('status', filters.status)
    if (filters?.profile_code) params.set('profile_code', filters.profile_code)
    if (filters?.issued_via) params.set('issued_via', filters.issued_via)
    if (filters?.valid !== undefined) params.set('valid', String(filters.valid))
    const qs = params.toString()
    return apiClient.get<CertificateSummary[]>(`/api/v1/certificates${qs ? `?${qs}` : ''}`)
  },
  issue: (payload: { csr_pem: string; profile_code: string; validity_days?: number }) =>
    apiClient.post<IssueCertificateResponse>('/api/v1/certificates', payload),
  revoke: (id: number, reason: string) =>
    apiClient.post<CertificateSummary>(`/api/v1/certificates/${id}/revoke`, { reason }),
  renew: (id: number, payload: { csr_pem: string; validity_days?: number }) =>
    apiClient.post<RenewCertificateResponse>(`/api/v1/certificates/${id}/renew`, payload),
}

File: backend/src/capki/db/models/certificates.py
import datetime as dt
import enum

from sqlalchemy import JSON, Enum, ForeignKey, Integer, String, Text
from sqlalchemy.orm import Mapped, mapped_column

from capki.db.base import Base, TimestampMixin, UTCDateTime


class CertificateStatus(str, enum.Enum):
    VALID = "valid"
    REVOKED = "revoked"
    EXPIRED = "expired"


class IssuedVia(str, enum.Enum):
    UI = "ui"
    API = "api"


class Certificate(TimestampMixin, Base):
    __tablename__ = "certificates"

    id: Mapped[int] = mapped_column(Integer, primary_key=True)
    ca_id: Mapped[int] = mapped_column(ForeignKey("certificate_authorities.id"), index=True)
    serial_hex: Mapped[str] = mapped_column(String(32), unique=True, index=True)
    profile_code: Mapped[str] = mapped_column(ForeignKey("cert_profiles.code"))
    subject_dn: Mapped[str] = mapped_column(String(500))
    sans: Mapped[list[str]] = mapped_column(JSON, default=list)
    certificate_pem: Mapped[str] = mapped_column(Text)
    csr_pem: Mapped[str] = mapped_column(Text)
    public_key_fingerprint: Mapped[str | None] = mapped_column(String(128), nullable=True)
    not_before: Mapped[dt.datetime] = mapped_column(UTCDateTime())
    not_after: Mapped[dt.datetime] = mapped_column(UTCDateTime())
    status: Mapped[CertificateStatus] = mapped_column(
        Enum(CertificateStatus, native_enum=False), default=CertificateStatus.VALID
    )
    requested_by_user_id: Mapped[int | None] = mapped_column(ForeignKey("users.id"), nullable=True)
    issued_via: Mapped[IssuedVia] = mapped_column(Enum(IssuedVia, native_enum=False))
    api_token_id: Mapped[int | None] = mapped_column(ForeignKey("api_tokens.id"), nullable=True)
    expiry_notified_at: Mapped[dt.datetime | None] = mapped_column(UTCDateTime(), nullable=True)


class Revocation(Base):
    __tablename__ = "revocations"

    id: Mapped[int] = mapped_column(Integer, primary_key=True)
    certificate_id: Mapped[int] = mapped_column(ForeignKey("certificates.id"), unique=True)
    revoked_at: Mapped[dt.datetime] = mapped_column(UTCDateTime())
    reason: Mapped[str] = mapped_column(String(32))
    revoked_by_user_id: Mapped[int | None] = mapped_column(ForeignKey("users.id"), nullable=True)


class CrlIssuance(Base):
    __tablename__ = "crl_issuances"

    id: Mapped[int] = mapped_column(Integer, primary_key=True)
    ca_id: Mapped[int] = mapped_column(ForeignKey("certificate_authorities.id"), index=True)
    crl_number: Mapped[int] = mapped_column(Integer)
    crl_pem: Mapped[str] = mapped_column(Text)
    this_update: Mapped[dt.datetime] = mapped_column(UTCDateTime())
    next_update: Mapped[dt.datetime] = mapped_column(UTCDateTime())
    created_at: Mapped[dt.datetime] = mapped_column(UTCDateTime())
